### PR TITLE
feat(frontend): Grouping util for token list

### PR DESCRIPTION
**Pull request by AntonioVentilii-DFINITY**
# Motivation

We want to change the way we group tokens, to have more than two elements in the group, and to try and avoid repetitions/glitches when the order of the token list is not the usual one.
Plus, the 

# Definitions

### Main Token

The "original" token of a certain group or of a certain other token; for example `BTC` is the "main token" of `ckBTC`, or `ETH` is the "main token" of `SepoliaETH`. Typically, a token and its "main token" should be 1:1.

### Secondary token

A token that has a "main token", even if not present in the list. It is just to indicate that it should always be grouped if the latter is present.

# Implementation

Following PR #3296 ,tThe function loops through the tokens list and groups them according to a prop key that links a token to its "main token". For example, it could be `twinToken` for ck-tokens. But the logic can be extended to other props, if necessary.

The tokens with no "main token" will still be included in a group, but it will be a single-element group, where the the token itself takes the place of the "main token". That, in general, makes sense for:
- tokens that are not "secondary tokens" of a "main token";
- tokens that have a "main token" but it is not present in the list.

The returned list respects the sorting order of the initial tokens list, meaning that the group is created at each position of the first encountered token of the group. So, independently of being a "main token" or a "secondary token", the group will replace the first token of the group in the list. That is useful if a "secondary token" is before the "main token" in the list; for example, if the list is sorted by balance.

# Tests

I created a few tests mocking some twin tokens and specific scenarios: different sorting, multiple "secondary tokens", missing "main token".

### CODE REVIEW

**Comment by AntonioVentilii-DFINITY** [comment]
[OISY-164]

**Comment by AntonioVentilii-DFINITY** [comment]
@peterpeterparker I think ready for another review

**Comment by AntonioVentilii-DFINITY** [comment]
@peterpeterparker ready for another review

File: src/frontend/src/lib/utils/token-group.utils.ts [hunk 1]
```diff
@@ -164,3 +165,52 @@ export const groupMainToken = ({ token, tokenGroup }: GroupTokenParams): TokenUi
...
+ */
+export const groupTokens = (tokens: TokenUi[]): TokenUiGroup[] => {
+	const tokenGroupsMap = tokens.reduce<{
+		[id: TokenId]: TokenUiGroup | undefined;
+	}>(
+		(acc, token) => ({
+			...acc,
+			...('twinToken' in token &&
+			isToken(token.twinToken) &&
+			token.decimals === token.twinToken.decimals
```
Comment (peterpeterparker) [inline src/frontend/src/lib/utils/token-group.utils.ts:195]: Not necesseraly for this PR but the comparison of decimals feels like something important so maybe worth a standalone function for it?

File: src/frontend/src/lib/utils/token.utils.ts [hunk 2]
```diff
@@ -183,3 +185,6 @@ export const sumUsdBalances = ([usdBalance1, usdBalance2]: [
  */
 export const isRequiredTokenWithLinkedData = (token: Token): token is RequiredTokenWithLinkedData =>
 	'twinTokenSymbol' in token && typeof token.twinTokenSymbol === 'string';
+
+export const isToken = (token: Option<unknown>): token is Token =>
```
Comment (peterpeterparker) [inline src/frontend/src/lib/utils/token.utils.ts:189]: Should we add a test for this?

**Review by peterpeterparker — APPROVED** [review]
LGTM, thx

**Comment by AntonioVentilii** [comment]
Superseded by:

-

**Pull request closed without merging**